Rewrite rule that simplifies predicates in binary logical expressions

sopel39 opened the pull request:
It is possible that common sub-predicate occures in multiple predicates.
For instance:
 (X AND Y) OR (X AND V)
or
 (X OR Y) AND (X OR V)

In such cases it is possible to extract this common predicate:
 X AND (Y OR V)
or
 X OR (Y AND V)

This will let join predicates to be pushed to the top, therefore
allowing for faster joins.

Additionally, this rewrite will remove duplicate predicates improving
query execution performance.

sopel39 commented:
This solves issue: https://github.com/facebook/presto/issues/4406

FYI: @martint @kokosing @ilfrin

losipiuk commented on presto-main/src/main/java/com/facebook/presto/sql/ExpressionUtils.java:
```diff
@@ -173,6 +185,18 @@ public static Expression stripDeterministicConjuncts(Expression expression)
+            case OR:
+                return LogicalBinaryExpression.Type.AND;
+            default:
+                throw new IllegalArgumentException("Unsupported comparison: " + type);
```
s/comparisson/ logical expression type/

losipiuk commented:
This optimizer just changes expression in FilterNode. Which optimizer is responsible for pushing new conjuncts down to JoinNode as `EquiJoinClause`s?

electrum commented:
@losipiuk I think you want `PredicatePushDown` which was mostly written by @erichwang

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/ExpressionUtils.java:
```diff
@@ -114,6 +111,21 @@ public static Expression binaryExpression(LogicalBinaryExpression.Type type, Ite
+        if (type == LogicalBinaryExpression.Type.AND) {
+            return combineConjuncts(expressions);
+        }
+        else {
```
Nit: The `else` is redundant

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPredicates.java:
```diff
@@ -0,0 +1,143 @@
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+public class SimplifyPredicates
```
Change name, that would say what it does

sopel39 replied:
well, it does exactly that in general. Specifically it does remove duplicate predicates and pushes up common predicates. Hard to give it a compact name. What name do you suggest?

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSimplifyPredicates.java:
```diff
@@ -0,0 +1,74 @@
+
+        assertSimplifies("((X OR V) AND V) OR ((X OR V) AND V)", "(X OR V) AND V");
+        assertSimplifies("((X OR V) AND X) OR ((X OR V) AND V)", "X OR V");
+        assertSimplifies("((X OR V) AND Z) OR ((X OR V) AND V)", "V OR (X AND Z)");
```
Nice test!

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPredicates.java:
```diff
@@ -0,0 +1,143 @@
+        return SimplePlanRewriter.rewriteWith(new Rewriter(), plan);
+    }
+
+    public static Expression simplifyPredicates(Expression expression)
```
This is only for test, so please move it there

sopel39 replied:
done

sopel39 replied:
It is not only used in tests. I could make it package private though. I am not sure it complies with Presto coding standard.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPredicates.java:
```diff
@@ -0,0 +1,143 @@
+
+        private List<List<Expression>> getSubPredicates(Iterable<Expression> predicates)
+        {
+            List<List<Expression>> subPredicates = new LinkedList<>();
```
Use ImmutableList.Builder

sopel39 replied:
done

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSimplifyPredicates.java:
```diff
@@ -0,0 +1,74 @@
+        assertSimplifies("(X OR Y OR V) AND (X OR Y OR Z)", "(X OR Y) OR (V AND Z)");
+
+        assertSimplifies("((X OR V) AND V) OR ((X OR V) AND V)", "(X OR V) AND V");
+        assertSimplifies("((X OR V) AND X) OR ((X OR V) AND V)", "X OR V");
```
Can you add a tests with nested common parts like
`(x and ((y and z) or (y and v)) or (x and y)`
To see how `y` will be handled.

kokosing replied:
And one more nested example
`(a and b and c and d) or (a and b and e) or (a and f)`

Add also a test without common parts to see that is unchanged

sopel39 replied:
subexpression were rewritten but not taken at upper level. nice catch.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/ExpressionUtils.java:
```diff
@@ -173,6 +185,18 @@ public static Expression stripDeterministicConjuncts(Expression expression)
                 .collect(toImmutableList()));
     }
 
+    public static LogicalBinaryExpression.Type flipLogicalBinaryExpressionType(LogicalBinaryExpression.Type type)
```
Maybe this method could be moved to LogicalBinatyExpression.Type enum?

losipiuk replied:
That class is generated by ANTLR

sopel39 replied:
It is similar to `flipComparison` method, so it's better to be consistent I think

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPredicates.java:
```diff
@@ -0,0 +1,143 @@
+                    .collect(Collectors.toList());
+
+            Expression result = combinePredicates(flippedNodeType,
+                    combinePredicates(flippedNodeType, commonPredicates),
```
There is no need to wrap common predicates in parenthesis as outer logical expession is of the same type

sopel39 replied:
combinePredicates accepts Expression... array. There is no `combinePredicates(type, Expression expression, List<Expression> rest)`. Both arguments need to be `Expression` (not `Expression` and `List<Expression>`)

kokosing commented:
Some comments, but generally it looks nice

kokosing commented:
It could be also applied for expressions in project or table scan nodes.

sopel39 commented:
fixed comments

sopel39 commented:
I have added a predicate simplifier phase that pushes down negations
to the bottom of expression tree. Added flip(), negate() methods 
to ComparisionExpression.Type and LogicalBinaryExpression.Type. 
Using predicates simplifier instead of ExpressionUtils.normalize().

Please have a look again. I will probably squash those commits when they pass review.
FYI: @losipiuk @kokosing @electrum

martint commented:
@erichwang can you review?

kokosing commented:
looks good to me

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPredicates.java:
```diff
@@ -0,0 +1,215 @@
+            else if (node.getValue() instanceof ComparisonExpression) {
+                ComparisonExpression child = (ComparisonExpression) node.getValue();
+                return new ComparisonExpression(
+                        child.getType().negate(),
```
i think you need to custom handle the case where the type is IS_DISTINCT_FROM. The negation of that type needs to handle the NULL value properly.

erichwang replied:
you should probably add a test case for this too

sopel39 replied:
I don't think we can negate IS_DISTINCT_FROM. I think we should just skip ComparisonExpression in such case. Added test

erichwang commented:
in general this looks pretty good.

one point to consider is that in the planner, we generally prefer to deal with conjunctions at the top level since those are easier to decompose and move around. So we may not always want to do the full normalization.

sopel39 commented:
Yes, the idea is that it will extract common predicates and make conjunction like in Tpch Q19. It could also make disjunction if there is a common predicate. It could help expression simplification higher up in the tree. If we end up with disjunction at the top it means that sub expreessions were already complex and join expressions were impossible (ORed expressions) to extract anyway if that is your concern.

erichwang commented:
I guess not just focusing on the join expressions, but predicate move around in general requires conjunctions. For example if you are doing a JOIN between table A and Table B and you had a predicate:
(a1 OR a2) AND (a1 OR b1)
then you could push (a1 OR a2) down to the A side, but in the normalized form, it is not as easily apparent.

sopel39 commented:
That makes sense. I have rebased on master, moved predicates simplification to SimplifyExpression and added condition that predicates should not be simplified for the top expression. I will squash after you review.

The next step would would be probably to use simplifier instead of ExpressionUtil.normalize, because currently if you write duplicate predicates in `on` or `using` join, it won't be simplified. The problem is though that analysis class is using IdentityMap for expression types and simplifier creates new, but equal nodes. This will have to be refactored before we can use simplifier.

erichwang commented:
last couple comments here =)

sopel39 commented:
rebased, applied comments and squashed

erichwang commented:
looks good

erichwang commented:
Merged, thanks!

haozhun commented:
We have identified a correctness issue introduced by this pull request. A more detailed explanation will be posted later.

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/ExpressionUtils.java:
```diff
@@ -114,6 +109,20 @@ public static Expression binaryExpression(LogicalBinaryExpression.Type type, Ite
+            return combineConjuncts(expressions);
+        }
+
+        return combineDisjuncts(expressions);
```
Default for combining disjuncts should be false?
